### PR TITLE
ENH: Add the split strategy to regression class

### LINKED ISSUE #297: [ENHANCEMENT] Split for regression & classification

### PR DESCRIPTION
**Pull request by thibaultcordier**
# Description

To make split conformal predictions, you have to pre-train a model on `X_train`, then use the `fit` method with the `cv="prefit"` parameter with `MapieRegressor` on `X_calib`. This PR proposes to do it automatically when using the `fit` method with the `cv="split"` parameter on `X = X_train U X_calib` (both on regression and classification when it is necessary).

Fix #297

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

- [x] Modify tests in `test_regression.py` file to verify the changes when using `cv="split"` (in particular: coverage, width).
- [x] Check if other tests need to be modified.

# Checklist

- [x] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
- [x] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
- [x] Linting passes successfully : `make lint`
- [x] Typing passes successfully : `make type-check`
- [x] Unit tests pass successfully : `make tests`
- [x] Coverage is 100% : `make coverage`
- [x] Documentation builds successfully : `make doc`

### CODE REVIEW

File: HISTORY.rst [hunk 1]
```diff
@@ -12,21 +12,27 @@ History
 
 0.6.2 (2023-03-22)
 ------------------
+
```
Comment (vincentblot28) [inline HISTORY.rst:15]: This file is not up to date, please pull master branch

File: mapie/regression.py [hunk 2]
```diff
@@ -275,7 +277,7 @@ def _check_agg_function(
                 "You need to specify an aggregation function when "
                 f"cv's type is in {self.cv_need_agg_function}."
             )
-        if (agg_function is not None) or (self.cv == "prefit"):
+        if (agg_function is not None) or (self.cv in ["prefit", "split"]):
```
Comment (vincentblot28) [inline mapie/regression.py:280]: Define `["prefit", "split"]` as a class attribute

File: mapie/regression.py [hunk 3]
```diff
@@ -424,12 +425,12 @@ def _aggregate_with_mask(self, x: NDArray, k: NDArray) -> NDArray:
...
         ArrayLike of shape (n_samples_test,)
-            Array of aggregated predictions for each testing  sample.
+            Array of aggregated predictions for each testing sample.
         """
-        if self.cv == "prefit":
+        if self.cv in ["prefit", "split"]:
             raise ValueError(
                 "There should not be aggregation of predictions if cv is "
-                "'prefit'"
+                "'prefit' or 'split'."
```
Comment (vincentblot28) [inline mapie/regression.py:433]: Idem, use a class attribute

Comment (thibaultcordier) [reply mapie/regression.py:433]: I propose to use "self.no_agg_cv_" class attribute. I propose also to do the same with "self.no_agg_methods_ = ['naive', 'base']". I am doing these changes.

File: HISTORY.rst (hunk 1 above)
Comment (thibaultcordier) [reply HISTORY.rst:23]: Ok, it is updated now.

File: mapie/utils.py [hunk 4]
```diff
@@ -160,15 +161,17 @@ def check_cv(
...
-    if (
-        isinstance(cv, BaseCrossValidator)
-        or (cv == "prefit")
-        or (cv == "split")
-    ):
+    if isinstance(cv, BaseCrossValidator):
+        return cv
+    if isinstance(cv, BaseShuffleSplit):
+        return cv
+    if cv in ["prefit"]:
```
Comment (vincentblot28) [inline mapie/utils.py:168]: Why using in List instead of == str ?

Comment (thibaultcordier) [reply mapie/utils.py:168]: There is no particular reason. I can change that with a string comparison if you think it is more pertinent. Let me know in order to add this change.

File: mapie/regression.py (hunk 2 above)
Comment (thibaultcordier) [reply mapie/regression.py]: I propose to use "self.no_agg_cv_" class attribute.

File: mapie/tests/test_common.py [hunk 5]
```diff
@@ -158,7 +158,7 @@ def test_invalid_method(MapieEstimator: BaseEstimator, method: str) -> None:
 
 @pytest.mark.parametrize("MapieEstimator", MapieSimpleEstimators())
 @pytest.mark.parametrize(
-    "cv", [-3.14, -2, 0, 1, "cv", LinearRegression(), [1, 2]]
+    "cv", [-3.14, -2, 1, 0, "cv", LinearRegression(), [2, 3]]
```
Comment (vincentblot28) [inline mapie/tests/test_common.py:161]: why this change ?

File: mapie/regression.py [hunk 6]
```diff
@@ -653,7 +655,8 @@ def predict(
 
         alpha_np = cast(NDArray, alpha)
         check_alpha_and_n_samples(alpha_np, n)
-        if self.method in ["naive", "base"] or self.cv == "prefit":
+
+        if self.method in ["naive", "base"] or self.cv in ["prefit", "split"]:
```
[the remark this answers is not in the document]
Comment (thibaultcordier) [reply mapie/regression.py]: I propose to use "self.no_agg_methods_" class attribute.

File: mapie/tests/test_common.py (hunk 5 above)
Comment (thibaultcordier) [reply mapie/tests/test_common.py:161]: Oh! I think it is a modification I made when testing my changes. It is no pertinent indeed.

File: mapie/tests/test_classification.py [hunk 7]
```diff
@@ -681,7 +682,10 @@ def test_valid_method(method: str) -> None:
...
 
 
-@pytest.mark.parametrize("cv", [None, -1, 2, KFold(), LeaveOneOut()])
+@pytest.mark.parametrize(
+    "cv", [None, -1, 2, KFold(), LeaveOneOut(),
+           ShuffleSplit(n_splits=1, test_size=0.1), "prefit"]
+)
 def test_valid_cv(cv: Any) -> None:
     """Test that valid cv raises no errors."""
     model = LogisticRegression(multi_class="multinomial")
```
Comment (vincentblot28) [inline mapie/tests/test_classification.py:691]: Why no test is implemented to test the result ? (like you did for regression)

File: mapie/calibration.py [hunk 8]
```diff
@@ -134,6 +134,36 @@ def __init__(
...
+        -------
+        str
+            'prefit' or 'split'.
+
+        Raises
+        ------
+        ValueError
+            If the cross-validator is not valid.
+        """
+        if cv in ["prefit", "split"]:
```
Comment (vincentblot28) [inline mapie/calibration.py:160]: Use class attribute

Comment (thibaultcordier) [reply mapie/calibration.py:160]: I propose "valid_cv" as class attribute.

File: mapie/regression.py [hunk 9]
```diff
@@ -424,12 +428,14 @@ def _aggregate_with_mask(self, x: NDArray, k: NDArray) -> NDArray:
...
         """
-        if self.cv == "prefit":
+        if self.method in self.no_agg_methods_ \
+                or self.cv in self.no_agg_cv_:
             raise ValueError(
-                "There should not be aggregation of predictions if cv is "
-                "'prefit'"
+                "There should not be aggregation of predictions"
+                "if cv is in 'self.no_agg_cv_'"
+                "or if method is in 'self.no_agg_methods_'."
```
Comment (vincentblot28) [inline mapie/regression.py:438]: idem

File: mapie/utils.py (hunk 4 above)
Comment (vincentblot28) [reply mapie/utils.py:168]: I think `==` is better

File: mapie/tests/test_common.py (hunk 5 above)
Comment (vincentblot28) [reply mapie/tests/test_common.py:161]: Ok, let's come back to the original

File: mapie/regression.py (hunk 9 above)
Comment (thibaultcordier) [reply mapie/regression.py:438]: Done.

File: mapie/tests/test_classification.py (hunk 7 above)
Comment (vincentblot28) [reply mapie/tests/test_classification.py:691]: ok, let's add them

Comment (vincentblot28) [reply mapie/tests/test_classification.py:691]: It is done in classification with this test `test_toy_dataset_predictions` which loops over strategies

File: mapie/calibration.py [hunk 10]
```diff
@@ -134,6 +136,36 @@ def __init__(
...
+        Raises
+        ------
+        ValueError
+            If the cross-validator is not valid.
+        """
+        if cv in self.valid_cv:
+            return cv
+        raise ValueError(
+            "Invalid cv argument. "
+            "Allowed values are 'prefit' or 'split'."
```
Comment (vincentblot28) [inline mapie/calibration.py:166]: Use class attribute

File: mapie/regression.py [hunk 11]
```diff
@@ -698,17 +711,17 @@ def predict(
...
                 for _alpha in alpha_np
             ]
-        ).data
+        )
         y_pred_up = np.column_stack(
             [
                 np_nanquantile(
                     upper_bounds.astype(float),
-                    1 - _alpha,
+                    1-_alpha,
```
Comment (vincentblot28) [inline mapie/regression.py:719]: wring linting, there should be a space before and after `-`

File: mapie/tests/test_calibration.py [hunk 12]
```diff
@@ -157,13 +157,30 @@ def test_other_methods() -> None:
         mapie_cal.fit(X, y)
 
 
-def test_prefit() -> None:
+def test_prefit_cv_argument() -> None:
     """Test that prefit method works"""
     est = RandomForestClassifier().fit(X, y)
     mapie_cal = MapieCalibrator(estimator=est, cv="prefit")
     mapie_cal.fit(X, y)
```
Comment (vincentblot28) [inline mapie/tests/test_calibration.py:166]: You should also test results with `cv=split`

File: mapie/calibration.py (hunk 10 above)
Comment (thibaultcordier) [reply mapie/calibration.py:166]: Done.

File: mapie/regression.py (hunk 11 above)
Comment (thibaultcordier) [reply mapie/regression.py:719]: Done.

File: mapie/tests/test_calibration.py (hunk 12 above)
Comment (thibaultcordier) [reply mapie/tests/test_calibration.py:166]: It is tested just after this test. I propose two tests because we nedd to prefit a model with "prefit" and it is not the case with "split".

Comment (vincentblot28) [reply mapie/tests/test_calibration.py:166]: yes tested that the model can be fitted, you didn't tested the results of the calibration, see `test_correct_results` test

File: mapie/tests/test_common.py (hunk 5 above)
Comment (thibaultcordier) [reply mapie/tests/test_common.py]: Done.

File: mapie/tests/test_calibration.py (hunk 12 above)
Comment (thibaultcordier) [reply mapie/tests/test_calibration.py:166]: Ok, I haven't understand that. Okay to add result test with `cv=split`

Comment (thibaultcordier) [reply mapie/tests/test_calibration.py:197]: Done technically.

File: mapie/tests/test_classification.py (hunk 7 above)
Comment (thibaultcordier) [reply mapie/tests/test_classification.py:691]: It was more difficult than expected: it was necessary to review the management of the random_state with the cv method (in connection with the Issue #290). To answer your request, I solved the problem directly but it will be necessary to complete the code with additional tests (future PR of Issue #290). Therefore, I added more changes than expected.

File: mapie/tests/test_calibration.py (hunk 12 above)
Comment (thibaultcordier) [reply mapie/tests/test_calibration.py:198]: Done.

File: mapie/tests/test_classification.py (hunk 7 above)
Comment (thibaultcordier) [reply mapie/tests/test_classification.py:735]: I have detected a side-effect with `cv=split`. I have made a correction. Now, the `test_toy_dataset_predictions` is also done for `cv=split`.

File: mapie/classification.py [hunk 13]
```diff
@@ -104,7 +109,7 @@ class MapieClassifier(BaseEstimator, ClassifierMixin):
         for evaluation quantiles and prediction sets in cumulated_score.
         Pass an int for reproducible output across multiple function calls.
 
-        By default ```1``.
+        By default ```None``.
```
Comment (vincentblot28) [inline mapie/classification.py:112]: why changin 1 to None ?
mission ```

Comment (thibaultcordier) [reply mapie/classification.py:112]: I have fixed ``` for classification and regression.

File: mapie/regression.py [hunk 14]
```diff
@@ -69,6 +68,9 @@ class MapieRegressor(BaseEstimator, RegressorMixin):
           - ``sklearn.model_selection.LeaveOneOut`` (jackknife),
           - ``sklearn.model_selection.KFold`` (cross-validation),
           - ``subsample.Subsample`` object (bootstrap).
+        - ``"split"``, does not involve cross-validation but a division
```
Comment (vincentblot28) [inline mapie/regression.py:71]: remove ""

File: mapie/regression.py [hunk 15]
```diff
@@ -112,7 +114,7 @@ class MapieRegressor(BaseEstimator, RegressorMixin):
         Jackknife+-after-Bootstrap method), this function is also used to
         aggregate the training set in-sample predictions.
 
-        If cv is ``"prefit"``, ``agg_function`` is ignored.
+        If cv is ``"prefit"`` or ``"split"``, ``agg_function`` is ignored.
```
Comment (vincentblot28) [inline mapie/regression.py:117]: remove ""

File: mapie/classification.py (hunk 13 above)
Comment (thibaultcordier) [reply mapie/classification.py:112]: Why changing 1 to None? Because by default, we use `None` (see line 194).

File: mapie/regression.py [hunk 16]
```diff
@@ -178,23 +187,28 @@ class MapieRegressor(BaseEstimator, RegressorMixin):
...
+    >>> mapie_reg = mapie_reg.fit(X_toy, y_toy)
     >>> y_pred, y_pis = mapie_reg.predict(X_toy, alpha=0.5)
     >>> print(y_pis[:, :, 0])
-    [[ 4.7972973   5.8       ]
-     [ 6.69767442  7.65540541]
-     [ 8.59883721  9.58108108]
-     [10.5        11.40116279]
-     [12.4        13.30232558]
-     [14.25       15.20348837]]
+    [[ 4.95714286  5.61428571]
```
Comment (vincentblot28) [inline mapie/regression.py:195]: why the results are different ?

File: mapie/regression.py (hunk 14 above)
Comment (thibaultcordier) [reply mapie/regression.py:71]: Why removing "" if we do the same for '"prefit"` (line 74)?

File: mapie/regression.py (hunk 16 above)
Comment (thibaultcordier) [reply mapie/regression.py:195]: Because before, "cv" methods haven't got `shuffle=True` when using `split` method. I change that.

Comment (thibaultcordier) [reply mapie/regression.py:195]: I did that because I have side-effect when using `split`.

File: mapie/tests/test_classification.py [hunk 17]
```diff
@@ -267,30 +278,31 @@
...
-    "cumulated_score_not_include_cv_mean": 5 / 9,
-    "cumulated_score_randomized_cv_mean": 5 / 9,
-    "cumulated_score_include_cv_crossval": 2 / 9,
-    "cumulated_score_not_include_cv_crossval": 0,
-    "cumulated_score_randomized_cv_crossval": 4 / 9,
-    "naive": 5 / 9,
-    "top_k": 1,
-    "raps": 1,
-    "raps_randomized": 8/9
+    'score': 0.6666666666666666,
```
Comment (vincentblot28) [inline mapie/tests/test_classification.py:281]: Why changing the format ? Moreover, to make the review easier please change only what is needed to be changes (typically, the results for non cross-validation methods should not be changed)

File: mapie/tests/test_classification.py [hunk 18]
```diff
@@ -267,30 +278,31 @@
...
+    'score_split': 0.8888888888888888,
+    'score_cv_mean': 1.0,
+    'score_cv_crossval': 1.0,
+    'cumulated_score_include': 1.0,
+    'cumulated_score_not_include': 0.5555555555555556,
```
Comment (vincentblot28) [inline mapie/tests/test_classification.py:286]: You should also test for the others method, to be as exhaustive as for the others functionnalities

File: mapie/tests/test_classification.py (hunk 17 above)
Comment (thibaultcordier) [reply mapie/tests/test_classification.py:281]: Because I have changed the parameter `shuffle=True` in `check_cv` function (see in `utils.py`), all the results are impacted. In fact, I do copy/paste. I agree to changing the format.

File: mapie/utils.py [hunk 19]
```diff
@@ -154,21 +162,29 @@ def check_cv(
         If the cross-validator is not valid.
     """
     if cv is None:
-        return KFold(n_splits=5)
+        return KFold(
```
[the remark this answers is not in the document]
Comment (thibaultcordier) [reply mapie/utils.py:165]: this is a proposition:

```suggestion
      if cv is None:
          return KFold(
              n_splits=5, shuffle=True, random_state=random_state
          )
      elif isinstance(cv, int):
          if cv == -1:
              return LeaveOneOut()
          elif cv >= 2:
              return KFold(
                  n_splits=cv, shuffle=True, random_state=random_state
              )
          else:
              raise ValueError(
                  "Invalid cv argument. "
                  "Allowed integer values are -1 or int >= 2."
              )
      elif isinstance(cv, BaseCrossValidator):
          return cv
      elif isinstance(cv, BaseShuffleSplit):
          return cv
      elif cv == "prefit":
          return cv
      elif cv == "split":
          return ShuffleSplit(
              n_splits=1, test_size=0.5, random_state=random_state
          )
      else:
          raise ValueError(
              "Invalid cv argument. "
              "Allowed values are None, -1, int >= 2, 'prefit', 'split', "
              "or a BaseCrossValidator object (Kfold, LeaveOneOut)."
          )
```

File: mapie/tests/test_classification.py (hunk 18 above)
Comment (thibaultcordier) [reply mapie/tests/test_classification.py:286]: I agree, but this is outside the current RP. I only propose the `split`cv method. But before this PR, `top_k`, `raps` and `raps_randomized` methods already exist without exhaustive tests (only on `prefit`). So that this PR is not too long, I propose to create a question to solve this problem.

Comment (thibaultcordier) [reply mapie/tests/test_classification.py]: Done.

File: mapie/utils.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (thibaultcordier) [reply mapie/utils.py:165]: `TypeError: '>=' not supported between instances of 'str' and 'int'`: side-effect if we remove `elif isinstance(cv, int)`

[the remark this answers is not in the document]
Comment (thibaultcordier) [reply mapie/utils.py:165]: Done.

File: mapie/tests/test_regression.py [hunk 20]
```diff
@@ -42,10 +43,31 @@
...
-    "jackknife_minmax": Params(method="minmax", agg_function="mean", cv=-1),
+    "naive": Params(
+        method="naive",
+        agg_function="median",
+        cv=None
+    ),
+    "split": Params(
+        method="base",
+        agg_function="median",
+        cv=ShuffleSplit(n_splits=1, test_size=0.1, random_state=1)
```
[the remark this answers is not in the document]
Comment (thibaultcordier) [reply mapie/tests/test_regression.py:54]: Because of the dependence of `random_state`

[the remark this answers is not in the document]
Comment (thibaultcordier) [reply mapie/tests/test_regression.py]: Done.

File: mapie/classification.py [hunk 21]
```diff
@@ -6,7 +6,8 @@
 import numpy as np
 from joblib import Parallel, delayed
 from sklearn.base import BaseEstimator, ClassifierMixin, clone
-from sklearn.model_selection import BaseCrossValidator, train_test_split
+from sklearn.model_selection import (BaseCrossValidator, train_test_split,
```
Comment (LacombeLouis) [inline mapie/classification.py:9]: Don't forget to use `isort` at the end to have the imports correctly ordered.

Comment (vincentblot28) [reply mapie/classification.py:9]: Louis has suffered so much from this type of comments that he is now a psycho just like me <3

Comment (vincentblot28) [reply mapie/classification.py:9]: So proud

Comment (thibaultcordier) [reply mapie/classification.py]: Done :)

File: mapie/tests/test_classification.py [hunk 22]
```diff
@@ -634,9 +678,10 @@ def test_default_parameters() -> None:
     assert mapie_clf.method == "score"
```
[the remark this answers is not in the document]
Comment (thibaultcordier) [reply mapie/tests/test_classification.py:728]: Yes, it is done.

File: mapie/tests/test_classification.py (hunk 18 above)
Comment (thibaultcordier) [reply mapie/tests/test_classification.py]: Okay, now it is done.

File: mapie/utils.py [hunk 23]
```diff
@@ -154,23 +162,37 @@ def check_cv(
...
-        "Allowed values are None, -1, int >= 2, 'prefit', "
-        "or a BaseCrossValidator object (Kfold, LeaveOneOut)."
-    )
+    elif isinstance(cv, BaseShuffleSplit):
+        return cv
+    elif cv == "prefit":
+        return cv
+    elif cv == "split":
+        return ShuffleSplit(
+            n_splits=1, test_size=0.5, random_state=random_state
```
Comment (vincentblot28) [inline mapie/utils.py:188]: Maybe it would be good to make the test_size a parameter of the function, and define the test_size as a class attribute, doing this, you able every one to change this ratio, it's a minor change but with a high added value

Comment (thibaultcordier) [reply mapie/utils.py:188]: I am not super fan to add a lot of parameters in class definition, but it makes sense. This change impacts also `MapieTimeSeriesRegressor`.

Comment (thibaultcordier) [reply mapie/utils.py:188]: Done.

File: mapie/utils.py [hunk 24]
```diff
@@ -128,11 +129,13 @@ def fit_estimator(
 
 
 def check_cv(
-    cv: Optional[Union[int, str, BaseCrossValidator]] = None
+    cv: Optional[Union[int, str, BaseCrossValidator]] = None,
+    test_size: Optional[Union[int, float]] = None,
```
Comment (vincentblot28) [inline mapie/utils.py:133]: test_size should only be a flaot

File: mapie/time_series_regression.py [hunk 25]
```diff
@@ -255,7 +264,8 @@ def predict(
         self.lower_quantiles_ = lower_quantiles
         self.higher_quantiles_ = higher_quantiles
 
-        if self.cv == "prefit":
+        if self.method in self.no_agg_methods_ \
+                or self.cv in self.no_agg_cv_:
```
Comment (vincentblot28) [inline mapie/time_series_regression.py:268]: no_agg_cv includes split, is it implemented here ?

File: mapie/regression.py [hunk 26]
```diff
@@ -209,18 +232,22 @@ def __init__(
         estimator: Optional[RegressorMixin] = None,
         method: str = "plus",
         cv: Optional[Union[int, str, BaseCrossValidator]] = None,
+        test_size: Optional[Union[int, float]] = None,
         n_jobs: Optional[int] = None,
         agg_function: Optional[str] = "mean",
         verbose: int = 0,
```
Comment (vincentblot28) [inline mapie/regression.py:238]: Move optional arguments at the end

File: mapie/time_series_regression.py (hunk 25 above)
Comment (thibaultcordier) [reply mapie/time_series_regression.py:268]: It inherits from `MapieRegressor`. So yes, it is implemented.

File: mapie/regression.py (hunk 26 above)
Comment (thibaultcordier) [reply mapie/regression.py:238]: All arguments are optional except `method` and `verbose`. Should I move `method` and `verbose` at the beginning?

File: mapie/classification.py [hunk 27]
```diff
@@ -1078,6 +1106,10 @@ def fit(
             self.conformity_scores_ = np.array([])
             self.k_ = np.empty_like(y_enc, dtype=int)
 
+        if self._target_type == "multiclass":
```
Comment (vincentblot28) [inline mapie/classification.py:1109]: This condition already exists

File: mapie/tests/test_classification.py [hunk 28]
```diff
@@ -267,21 +332,25 @@
...
+    "cumulated_score_include_split": 8/9,
+    "cumulated_score_not_include_split": 5/9,
+    "cumulated_score_randomized_split": 7/9,
+    "cumulated_score_include_cv_mean": 1.0,
+    "cumulated_score_not_include_cv_mean": 5/9,
+    "cumulated_score_randomized_cv_mean": 8/9,
+    "cumulated_score_include_cv_crossval": 4/9,
+    "cumulated_score_not_include_cv_crossval": 1/9,
+    "cumulated_score_randomized_cv_crossval": 7/9,
+    "naive": 5/9,
```
Comment (vincentblot28) [inline mapie/tests/test_classification.py:351]: add split for top_k, raps, and randomized_raps

File: mapie/classification.py (hunk 27 above)
Comment (thibaultcordier) [reply mapie/classification.py]: Done.

File: mapie/tests/test_classification.py (hunk 28 above)
Comment (thibaultcordier) [reply mapie/tests/test_classification.py:375]: Done.

**Comment by thibaultcordier** [comment]
@vincentblot28 @LacombeLouis, does this PR need further modification?

File: mapie/tests/test_classification.py (hunk 28 above)
Comment (vincentblot28) [reply mapie/tests/test_classification.py:375]: still missing for RAPS I think

Comment (thibaultcordier) [reply mapie/tests/test_classification.py:375]: I did not implement the split method for RAPS.

Comment (thibaultcordier) [reply mapie/tests/test_classification.py:399]: Done now.

File: mapie/tests/test_classification.py [hunk 29]
```diff
@@ -779,6 +992,38 @@ def test_y_is_list_of_string(
     assert y_ps.shape == (X.shape[0], len(np.unique(y)), n_alpha)
 
 
+def test_same_results_prefit_split() -> None:
```
Comment (vincentblot28) [inline mapie/tests/test_classification.py:995]: Could you just iterate over strategies ?

File: mapie/classification.py [hunk 30]
```diff
@@ -962,12 +985,15 @@ def fit(
 
         if self._target_type == "multiclass":
             if self.method == "raps":
-                X, self.X_raps, y_enc, self.y_raps = train_test_split(
-                    X,
-                    y_enc,
-                    test_size=size_raps,
-                    random_state=self.random_state
+                raps_split = ShuffleSplit(
+                    1, test_size=size_raps, random_state=self.random_state
```
Comment (vincentblot28) [inline mapie/classification.py:989]: why this change ?

File: mapie/tests/test_classification.py (hunk 29 above)
Comment (thibaultcordier) [reply mapie/tests/test_classification.py:995]: Over methods like "raps", "score", "cumulated_score" ?

File: mapie/classification.py (hunk 30 above)
Comment (thibaultcordier) [reply mapie/classification.py:990]: I need to have access to the index list (for train and and val split).

File: mapie/tests/test_classification.py (hunk 29 above)
Comment (thibaultcordier) [reply mapie/tests/test_classification.py]: Done.

**Review by vincentblot28 — APPROVED** [review]
